pallets/session-ext/src/manager.rs

```diff
@@ -0,0 +1,124 @@
+    traits::EraInfoProvider,
+};
+
+/// We assume that block `B` ends session nr `S`, and current era index is `E`.
+///
+/// 1. Block `B` initialized
+/// 2. `end_session(S)` is called
+/// *  Based on block count we might mark the session for a given validator as underperformed
+/// *  We update rewards and clear block count for the session `S`.
+/// 3. `start_session(S + 1)` is called.
+/// *  if session `S+1` starts new era we populate totals and unban all validators whose ban expired.
+/// *  if session `S+1` % [`BanConfig::clean_session_counter_delay`] == 0, we
+///    clean up underperformed session counter
+/// 4. `new_session(S + 2)` is called.
+/// *  If session `S+2` starts new era we emit fresh bans events
+/// *  We rotate the validators for session `S + 2` using the information about reserved and non reserved validators.
+
+impl<T> pallet_authorship::EventHandler<T::AccountId, T::BlockNumber> for Pallet<T>
+where
+    T: Config,
+{
+    fn note_author(validator: T::AccountId) {
+        SessionValidatorBlockCount::<T>::mutate(&validator, |count| {
+            *count += 1;
+        });
+    }
+}
+
+/// SessionManager that also fires EraManager functions.
+/// The order of the calls are as follows:
+/// First call is always from the inner SessionManager then the call to EraManager fn if applicable.
+/// * New session is planned:
+/// 1. Inner T::new_session invoked
+/// 2. If session starts era EM::on_new_era invoked
+/// 3. Logic related to new session from this pallet is invoked
+/// * Session ends:
+/// 1. Inner T::end_session invoked
+/// 2. Logic related to new session from this pallet is invoked
+/// * Session starts:
+/// 1. Inner T::start_session invoked
+/// 2. Logic related to new session from this pallet is invoked
+/// 3. If session starts era EM::new_era_start invoked
+/// 4. If session starts era logic related to new era from this pallet is invoked
+pub struct SessionManagerExt<E, EM, T, C>(PhantomData<(E, EM, T, C)>)
+where
+    T: SessionManager<C::AccountId>,
+    EM: EraManager,
+    E: EraInfoProvider,
+    C: Config;
+
+impl<E, EM, T, C> SessionManagerExt<E, EM, T, C>
+where
+    T: SessionManager<C::AccountId>,
+    E: EraInfoProvider,
+    EM: EraManager,
+    C: Config,
+{
+    fn session_starts_era(session: SessionIndex, next_era: bool) -> Option<EraIndex> {
+        let mut active_era = match E::active_era() {
+            Some(ae) => ae,
+            // no active era, session can't start it
+            _ => return None,
+        };
+        if next_era {
+            active_era += 1;
+        }
+        if let Some(era_start_index) = E::era_start_session_index(active_era) {
+            return if era_start_index == session {
+                Some(active_era)
+            } else {
+                None
+            };
+        }
+
+        None
+    }
+}
+
+impl<E, EM, T, C> SessionManager<C::AccountId> for SessionManagerExt<E, EM, T, C>
+where
+    T: SessionManager<C::AccountId>,
```

Comment:
Please, sort bounds in the order of declaration.

Reply:
done